lib/filewatch/watch.rb

```diff
@@ -51,7 +51,10 @@ def subscribe(observer, sincedb_collection)
           glob = 0
         end
         break if quit?
+        # NOTE: maybe the plugin should validate stat_interval <= sincedb_write_interval <= sincedb_clean_after
         sleep(@settings.stat_interval)
+        # we need to check potential expired keys (sincedb_clean_after) periodically
+        sincedb_collection.flush_at_interval
```

Comment:
:thinking:

We already have a `SincedbCollection#write_if_requested` in this `until quit?` loop, which conditionally sends `SincedbCollection#flush_at_interval` IFF a write has been requested. Should we simply change this to send `SincedbCollection#flush_at_interval` instead, or is there a valid reason to sometimes flush twice in a loop?

Also, this isn't new to your PR, but it looks like we are passing an instance of `SincedbCollection` around a fair bit, with this `flush_at_interval` and other state-mutating methods being invoked by potentially multiple threads simultaneously 😩. We should probably file a separate issue to audit the thread-safety.

Reply:
as the described in the bug report and in the description the problem is when no writes are requested (no new discovered files or new content in watched files) `write_if_requested` is simply a noop and thus `sincedb_clean_after` won't get triggered.

the plugin should guarantee a cleanup of expired sincedb entried regardless of whether there's actual new content. currently cleanup is only guaranteed to happen on shutdown.

have decided to add an explicit `flush_at_interval` along side while keeping the existing `write_if_requested`, since they accomplish different things we want to happen in a loop a) write changes if any b) guarantee periodic sincedb flushes

Reply:
@yaauie is your preference still (given ^^^) : 
> I think that the change to FileWatch::Watch#subscribe can be simplified a bit by jumping straight to invoking SincedbCollection#flush_at_interval exactly once per loop instead of invoking it once-or-twice.

do not have strong opinions for either way given the current state of code.

Reply:
You're right, that the actual flush is safeguarded to not flush more often than the interval, so it doesn't matter much if we attempt it once or twice in the loop. I'll update my review to an LGTM.